### PR TITLE
[LLDB] Improve ObjectFileELF files ability to load from memory.

### PR DESCRIPTION
**Pull request by clayborg**
This is the first part in improving ELF core files in LLDB. This first patch teaches LLDB to get as much data as possible from an in memory ELF image where we don't have section headers mapped into the process' memory. Core files contain the main executable mapped into memory so this patch will allow us to read it in, get the UUID and also find the DT_DEBUG entry so the dynamic loader can get a full list of shared libraries from the linked list in memory.

This patch modified ObjectFileELF to be able to more successfully load ELF files from memory. Highlights include:
- Load the .dynamic section using the PT_DYNAMIC program header if there are no sections headers.
- Load ELF notes from the PT_NOTE program header if there are no section headers.
- Modify the ObjectFileELF::GetArchitecture() to use the PT_NOTE segment if no section headers are available.
- Modify the ObjectFileELF::GetUUID() to use the PT_NOTE segment if no section headers are available. This allows us to find the UUID of a binary in core files without having the executable in disk.
- Modify ObjectFileELF::GetSegmentData() to read data from memory if the program header's data doesn't exist in the first memory region for the object file in memory.
- Modify ObjectFileELF::GetImageInfoAddress() to not rely on section headers. This will allow us in a follow up patch to not require the executable when loading an ELF core file as we can load the executable image from memory and find the DT_DEBUG entry.
- Fix an issue where we would not check if we got any data when we tried to parse the section headers from memory and we would create many SHT_NULL sections. We now check if we have data before we resize our section_headers array.
- Modify cwObjectFileELF::ParseDynamicSymbols() to parse and cache all dynamic entries and their names and change all code that read dynamic entries over to using the cached values. Previously different areas of code would manually parse the dynamic entries and some would get the names, and some wouldn't.
- Add dumping the the .dynamic section when running the "target modules dump objfile" command. This is used to test that this features works in the unit test and verifies we can get the needed DT_* entries needed for this to work.
- Fix dumping of the program headers to work for 64 bit and have the table aligned to the ASCII column headers.
- Load .dynstr string table from the DT_STRTAB dynamic entry.
- Load the dynamic symbol table from the DT_SYMTAB dynamic entry.
- Modify the lldb_private::Process to use the memory region information to get the size the the memory region that contains the object file header to make sure we read as much as we can and don't run into loading the program headers, .dynstr or .dynsym from program headers

### CODE REVIEW

**Comment by clayborg** [comment]
If anyone knows any ELF file experts, please add them as reviewers

**Comment by MaskRay** [comment]
I'm not familiar with lldb, but I can make some comments as @tschuett invited me:)

Parsing PT_NOTE is a great step, as program headers are sufficient for executables, shared objects, and core dumps.
The section header table isn't needed and the relevant code could be dropped (unless relocatable file debugging is required).

Has the code been updated to handle p_align=8 PT_LOAD correctly? https://reviews.llvm.org/D150022 might be a related change from LLVMObject.

**Comment by tschuett** [comment]
I heard rumours that you know a bit about ELF :)

**Comment by labath** [comment]
I like this idea a lot, but I have some reservations about the implementation.

For one, I think this patch is too big. I once heard someone say "if you have bullet points in your patch description, then the patch is doing too much". While I don't think we should go as far as to create a separate PR for each of your bullet points, I splitting it up into a couple of pieces would go along way towards making it easier to review.

I also see that some of the functionality is guarded by `IsInMemory()`, which doesn't sounds like the right thing to do, as the we should in principle be able to use the same code for parsing parsing section-header-less object files on disk. These aren't exactly common, but like @MaskRay said, section headers aren't strictly needed for linked object files, and it's much easier to make test using these.

Finally, I think that structuring some of this code as "fallback" is not ideal, as it can cause some data can be parsed twice (I think it happens at least with ELF notes in this patch). Even if that's innocuous , I don't think it's right because the two mechanisms (program and section headers) are just different ways of finding the same data. I think it'd be cleaner if this was implemented as a two-step process:
1. find the data (e.g., notes): This will look into section and program headers to find the appropriate bytes (I might even argue it should look at program headers first, as that's what the operating system will use)
2. use the data (regardless of where it comes from)

I realise this feedback isn't very specific, but that's because I found it very hard to follow everything that's going on in this patch. I'm sure I'll be able to be more specific on the partial patches (and maybe some of my assumptions will turn out to be incorrect). As a first patch in the series, I'd recommend teaching lldb to parse section-header-less object files. Right now, if I run lldb on such a file, it will tell me that it's empty (has zero sections). Making the program headers visible would lay the foundation for other changes, and it would also be the smallest testable piece of functionality (by dumping the section list).

@MaskRay can you recommend a good to create these kinds of files? I was thinking of a combination `yaml2obj` + `llvm-objcopy --strip-sections` (because IIRC yaml2obj always inserts some sections into the output), but maybe there's something better (?)

**Comment by clayborg** [comment]
ok, I will break this up to make it easier to review.

**Comment by MaskRay** [comment]
> I like this idea a lot, but I have some reservations about the implementation.
> 
> For one, I think this patch is too big. I once heard someone say "if you have bullet points in your patch description, then the patch is doing too much". While I don't think we should go as far as to create a separate PR for each of your bullet points, I do believe that splitting it up into a couple of pieces would go along way towards making it easier to review.
> 
> I also see that some of the functionality is guarded by `IsInMemory()`, which doesn't sounds like the right thing to do, as the we should in principle be able to use the same code for parsing parsing section-header-less object files on disk. These aren't exactly common, but like @MaskRay said, section headers aren't strictly needed for linked object files, and it's much easier to make test using these.
> 
> Finally, I think that structuring some of this code as "fallback" is not ideal, as it can cause some data can be parsed twice (I think it happens at least with ELF notes in this patch). Even if that's innocuous , I don't think it's right because the two mechanisms (program and section headers) are just different ways of finding the same data. I think it'd be cleaner if this was implemented as a two-step process:
> 
> 1. find the data (e.g., notes): This will look into section and program headers to find the appropriate bytes (I might even argue it should look at program headers first, as that's what the operating system will use)
> 2. use the data (regardless of where it comes from)
> 
> I realise this feedback isn't very specific, but that's because I found it very hard to follow everything that's going on in this patch. I'm sure I'll be able to be more specific on the partial patches (and maybe some of my assumptions will turn out to be incorrect). As a first patch in the series, I'd recommend teaching lldb to parse section-header-less object files. Right now, if I run lldb on such a file, it will tell me that it's empty (has zero sections). Making the program headers visible would lay the foundation for other changes, and it would also be the smallest testable piece of functionality (by dumping the section list).
> 
> @MaskRay can you recommend a good to create these kinds of files? I was thinking of a combination `yaml2obj` + `llvm-objcopy --strip-sections` (because IIRC yaml2obj always inserts some sections into the output), but maybe there's something better (?)

yaml2obj omits the section header table when `NoHeaders: true` is specified.
```
  - Type: SectionHeaderTable
    NoHeaders: true
```

However, obj2yaml doesn't create `NoHeaders: true` yet. 

If the test utilizes `ld.lld`, `llvm-objcopy --strip-sections` will be needed. GNU ld 2.41 supports `-z nosectionheader`, which lld doesn't support yet (if there is enough interest I can add it).

**Comment by labath** [comment]
> > @MaskRay can you recommend a good to create these kinds of files? I was thinking of a combination `yaml2obj` + `llvm-objcopy --strip-sections` (because IIRC yaml2obj always inserts some sections into the output), but maybe there's something better (?)
> 
> yaml2obj omits the section header table when `NoHeaders: true` is specified.
> 
> ```
>   - Type: SectionHeaderTable
>     NoHeaders: true
> ```
> 
> However, obj2yaml doesn't create `NoHeaders: true` yet.
> 
> If the test utilizes `ld.lld`, `llvm-objcopy --strip-sections` will be needed. GNU ld 2.41 supports `-z nosectionheader`, which lld doesn't support yet (if there is enough interest I can add it).

Thanks. This is very helpful. For basic tests, I think we could use yaml2obj, but it would be also nice to have a test which builds and executes a real section-free executable. That one would use a linker, but we should be able to do it by just postprocessing the built executable with llvm-objcopy.

So, while we would most likely use this feature if it existed, I think we can get by without it.

**Comment by clayborg** [comment]
I am going to split this up. New PR with just .dynamic changes is here:

https://github.com/llvm/llvm-project/pull/101237
